app/javascript/followButtons/initializeFollowButtons.js

```diff
@@ -0,0 +1,358 @@
+import { getInstantClick } from '../topNavigation/utilities';
+
+/* global showLoginModal  userData */
+
+/**
+ * Sets the text content of the button to the correct 'Follow' state
+ *
+ * @param {HTMLElement} button The Follow button to update
+ * @param {string} style The style of the button from its "info" data attribute
+ */
+function addButtonFollowText(button, style) {
+  switch (style) {
+    case 'small':
+      button.textContent = '+';
+      break;
+    case 'follow-back':
+      button.textContent = 'Follow back';
+      break;
+    default:
+      button.textContent = 'Follow';
+  }
+}
+
+/**
+ * Sets the text content of the button to the correct 'Following' state
+ *
+ * @param {HTMLElement} button The Follow button to update
+ * @param {string} style The style of the button from its "info" data attribute
+ */
+function addButtonFollowingText(button, style) {
+  button.textContent = style === 'small' ? '✓' : 'Following';
+}
+
+/**
+ * Changes the visual appearance and 'verb' of the button to match the new state
+ *
+ * @param {HTMLElement} button The Follow button to be updated
+ */
+function optimisticallyUpdateButtonUI(button) {
+  const { verb: newState } = button.dataset;
+  const buttonInfo = JSON.parse(button.dataset.info);
+  const { style } = buttonInfo;
+
+  // Often there are multiple follow buttons for the same followable item on the page
+  // We collect all buttons which match the click, and update them all
+  const matchingFollowButtons = Array.from(
+    document.getElementsByClassName('follow-action-button'),
+  ).filter((button) => {
+    const { info } = button.dataset;
+    if (info) {
+      const { id } = JSON.parse(info);
```

Comment:
tbh I'm not too keen on this, as it means every time a user clicks a follow button, we parse the json 'info' on every other follow button on the page 😬  (this is how we were doing it before too).

I'm not totally sure why we have the "info" bundled into one data attribute (there are three items inside). I'm wondering if it's worth in this refactor bringing the userId at least out into its own data attribute, then it would be straightforward to use a `querySelector` to grab any follow buttons with the same userId without all of this json parsing 🤔

Reply:
Probably not a bad thing. Definitely worth looking at in a subsequent PR.